pkg/services/ngalert/schedule/schedule.go

```diff
@@ -737,16 +743,28 @@ func newAlertRuleInfo(parent context.Context) *alertRuleInfo {
 	return &alertRuleInfo{evalCh: make(chan *evaluation), updateCh: make(chan struct{}), ctx: ctx, stop: cancel}
 }
 
-// eval signals the rule evaluation routine to perform the evaluation of the rule. Does nothing if the loop is stopped
-func (a *alertRuleInfo) eval(t time.Time, version int64) bool {
+// eval signals the rule evaluation routine to perform the evaluation of the rule. Does nothing if the loop is stopped.
+// Before sending a message into the channel, it does non-blocking read to make sure that there is no concurrent send operation.
+// Returns a tuple where first element is
+//   - true when message was sent
+//   - false when the send operation is stopped
+// the second element contains a dropped message that was sent by a concurrent sender.
+func (a *alertRuleInfo) eval(t time.Time, version int64) (bool, *evaluation) {
+	// read the channel in unblocking manner to make sure that there is no concurrent send operation.
+	var droppedMsg *evaluation
```

Comment:
What happens if the rule's evaluation routine pulls from the channel a nanosecond before this select locks the channel? The routine will have another tick to evaluate immediately after, right? So you can get into a situation where you're evaluating twice in rapid succession, followed by a break before the next evaluation. Certainly better than we're currently doing, but could we do better? 

If we gave each routine a ticker we could close stop the ticker when the rule is cleaned up, which would be equivalent to the channel closing we're doing now. Then you don't even need to schedule goroutines to feed the channel of a rule, the scheduler just needs to manage initializing, resetting, or cleaning up rules.

Reply:
> What happens if the rule's evaluation routine pulls from the channel a nanosecond before this select locks the channel? The routine will have another tick to evaluate immediately after, right? So you can get into a situation where you're evaluating twice in rapid succession, followed by a break before the next evaluation. Certainly better than we're currently doing, but could we do better?

that is the desired behavior. It is ok because the time at which the rule should be evaluated is constant, i.e. the current tick in the channel is X and the next is always `X + interval`.

>If we gave each routine a ticker we could close stop the ticker when the rule is cleaned up, which would be equivalent to the channel closing we're doing now. Then you don't even need to schedule goroutines to feed the channel of a rule, the scheduler just needs to manage initializing, resetting, or cleaning up rules.

Currently, all rule evaluations are synchronized against a single heartbeat. The scheduler spreads evaluations of all rules in an interval of 10 seconds, i.e. each rule's evaluation gets delayed to a tiny amount of time. By doing that it spreads the load (data source, the host that runs Grafana).
For example, the scheduler will send the eval message `tick X`  to the evaluator of rule 1 at time X + 10ms and for alert rule 2 - at time X + 25ms, etc. In other words, the scheduler is in control of the evaluation of all rules. 
What you suggest is to give more autonomy to a single rule.  This would be a fundamental change in how alerting engine works, and it should be discussed and designed appropriately. Both approaches have pros and cons.